File: src/local_rag/parsers/calibre.py
"""Calibre metadata.db parser for extracting book metadata."""

import logging
import sqlite3
from dataclasses import dataclass, field
from pathlib import Path

from bs4 import BeautifulSoup

logger = logging.getLogger(__name__)


@dataclass
class CalibreBook:
    """A book entry from a Calibre library with full metadata."""

    book_id: int
    title: str
    authors: list[str]
    tags: list[str]
    series: str | None
    series_index: float | None
    publisher: str | None
    pubdate: str | None
    rating: int | None  # 0-10 from Calibre
    languages: list[str]
    identifiers: dict[str, str]  # {"isbn": "...", "amazon": "..."}
    description: str | None  # plain text (stripped from HTML)
    formats: dict[str, str]  # {"PDF": "filename_without_ext", ...}
    relative_path: str  # books.path: "Author/Title (ID)"
    last_modified: str


def parse_calibre_library(library_path: Path) -> list[CalibreBook]:
    """Load all books with metadata from a Calibre library.

    Opens metadata.db in read-only mode to avoid any writes or lock conflicts.

    Args:
        library_path: Path to the Calibre library root directory.

    Returns:
        List of CalibreBook objects with full metadata.
    """
    db_path = library_path / "metadata.db"
    if not db_path.exists():
        logger.error("Calibre metadata.db not found at: %s", db_path)
        return []

    uri = f"file:{db_path}?mode=ro"
    try:
        conn = sqlite3.connect(uri, uri=True)
        conn.row_factory = sqlite3.Row
    except sqlite3.OperationalError as e:
        logger.error("Failed to open Calibre database at %s: %s", db_path, e)
        return []

    try:
        return _load_all_books(conn)
    except Exception as e:
        logger.error("Failed to parse Calibre library at %s: %s", library_path, e)
        return []
    finally:
        conn.close()


def _load_all_books(conn: sqlite3.Connection) -> list[CalibreBook]:
    """Bulk-load all books with metadata from the database."""
    # Load base book data
    books_rows = conn.execute(
        "SELECT id, title, path, pubdate, last_modified FROM books"
    ).fetchall()

    if not books_rows:
        return []

    # Build lookup maps for related tables
    authors_map = _load_book_authors(conn)
    tags_map = _load_book_tags(conn)
    series_map = _load_book_series(conn)
    publishers_map = _load_book_publishers(conn)
    ratings_map = _load_book_ratings(conn)
    languages_map = _load_book_languages(conn)
    identifiers_map = _load_book_identifiers(conn)
    comments_map = _load_book_comments(conn)
    formats_map = _load_book_formats(conn)

    books: list[CalibreBook] = []
    for row in books_rows:
        book_id = row["id"]

        description_html = comments_map.get(book_id)
        description = _strip_html_description(description_html) if description_html else None

        series_info = series_map.get(book_id)

        books.append(
            CalibreBook(
                book_id=book_id,
                title=row["title"],
                authors=authors_map.get(book_id, []),
                tags=tags_map.get(book_id, []),
                series=series_info[0] if series_info else None,
                series_index=series_info[1] if series_info else None,
                publisher=publishers_map.get(book_id),
                pubdate=row["pubdate"],
                rating=ratings_map.get(book_id),
                languages=languages_map.get(book_id, []),
                identifiers=identifiers_map.get(book_id, {}),
                description=description,
                formats=formats_map.get(book_id, {}),
                relative_path=row["path"],
                last_modified=row["last_modified"],
            )
        )

    logger.info("Loaded %d books from Calibre library", len(books))
    return books


def _load_book_authors(conn: sqlite3.Connection) -> dict[int, list[str]]:
    """Load author names grouped by book ID."""
    result: dict[int, list[str]] = {}
    rows = conn.execute(
        """SELECT bal.book, a.name
           FROM books_authors_link bal
           JOIN authors a ON bal.author = a.id
           ORDER BY bal.book, bal.id"""
    ).fetchall()
    for row in rows:
        result.setdefault(row["book"], []).append(row["name"])
    return result


def _load_book_tags(conn: sqlite3.Connection) -> dict[int, list[str]]:
    """Load tags grouped by book ID."""
    result: dict[int, list[str]] = {}
    rows = conn.execute(
        """SELECT btl.book, t.name
           FROM books_tags_link btl
           JOIN tags t ON btl.tag = t.id"""
    ).fetchall()
    for row in rows:
        result.setdefault(row["book"], []).append(row["name"])
    return result


def _load_book_series(conn: sqlite3.Connection) -> dict[int, tuple[str, float | None]]:
    """Load series info grouped by book ID. Returns (series_name, series_index)."""
    result: dict[int, tuple[str, float | None]] = {}
    rows = conn.execute(
        """SELECT bsl.book, s.name, b.series_index
           FROM books_series_link bsl
           JOIN series s ON bsl.series = s.id
           JOIN books b ON bsl.book = b.id"""
    ).fetchall()
    for row in rows:
        result[row["book"]] = (row["name"], row["series_index"])
    return result


def _load_book_publishers(conn: sqlite3.Connection) -> dict[int, str]:
    """Load publisher names grouped by book ID."""
    result: dict[int, str] = {}
    rows = conn.execute(
        """SELECT bpl.book, p.name
           FROM books_publishers_link bpl
           JOIN publishers p ON bpl.publisher = p.id"""
    ).fetchall()
    for row in rows:
        result[row["book"]] = row["name"]
    return result


def _load_book_ratings(conn: sqlite3.Connection) -> dict[int, int]:
    """Load ratings grouped by book ID."""
    result: dict[int, int] = {}
    rows = conn.execute(
        """SELECT brl.book, r.rating
           FROM books_ratings_link brl
           JOIN ratings r ON brl.rating = r.id"""
    ).fetchall()
    for row in rows:
        result[row["book"]] = row["rating"]
    return result


def _load_book_languages(conn: sqlite3.Connection) -> dict[int, list[str]]:
    """Load languages grouped by book ID."""
    result: dict[int, list[str]] = {}
    rows = conn.execute(
        """SELECT bll.book, l.lang_code
           FROM books_languages_link bll
           JOIN languages l ON bll.lang_code = l.id"""
    ).fetchall()
    for row in rows:
        result.setdefault(row["book"], []).append(row["lang_code"])
    return result


def _load_book_identifiers(conn: sqlite3.Connection) -> dict[int, dict[str, str]]:
    """Load identifiers grouped by book ID."""
    result: dict[int, dict[str, str]] = {}
    rows = conn.execute(
        "SELECT book, type, val FROM identifiers"
    ).fetchall()
    for row in rows:
        result.setdefault(row["book"], {})[row["type"]] = row["val"]
    return result


def _load_book_comments(conn: sqlite3.Connection) -> dict[int, str]:
    """Load comments (descriptions) grouped by book ID."""
    result: dict[int, str] = {}
    rows = conn.execute(
        "SELECT book, text FROM comments"
    ).fetchall()
    for row in rows:
        result[row["book"]] = row["text"]
    return result


def _load_book_formats(conn: sqlite3.Connection) -> dict[int, dict[str, str]]:
    """Load available formats grouped by book ID. Returns {format: filename_without_ext}."""
    result: dict[int, dict[str, str]] = {}
    rows = conn.execute(
        "SELECT book, format, name FROM data"
    ).fetchall()
    for row in rows:
        result.setdefault(row["book"], {})[row["format"]] = row["name"]
    return result


def get_book_file_path(
    library_path: Path, book: CalibreBook, preferred_formats: list[str] | None = None
) -> tuple[Path, str] | None:
    """Resolve the absolute file path for the best available format of a book.

    Args:
        library_path: Root path of the Calibre library.
        book: The CalibreBook to find the file for.
        preferred_formats: Ordered list of preferred formats (e.g. ["EPUB", "PDF"]).
            Defaults to ["EPUB", "PDF"].

    Returns:
        (absolute_path, format_lower) tuple, or None if no supported file found.
    """
    if preferred_formats is None:
        preferred_formats = ["EPUB", "PDF"]

    for fmt in preferred_formats:
        filename_base = book.formats.get(fmt)
        if filename_base is None:
            continue
        file_path = library_path / book.relative_path / f"{filename_base}.{fmt.lower()}"
        if file_path.exists():
            return file_path, fmt.lower()

    return None


def _strip_html_description(html: str) -> str:
    """Convert an HTML book description/synopsis to plain text."""
    soup = BeautifulSoup(html, "html.parser")
    text = soup.get_text(separator="\n")
    lines = [line.strip() for line in text.splitlines()]
    return "\n".join(line for line in lines if line)

File: internal/parser/calibre.go
package parser

import (
	"database/sql"
	"fmt"
	"log/slog"
	"path/filepath"
	"strings"

	_ "github.com/mattn/go-sqlite3"
)

// CalibreBook is a book entry from a Calibre library.
type CalibreBook struct {
	BookID       int64
	Title        string
	Authors      []string
	Tags         []string
	Series       string
	SeriesIndex  float64
	Publisher    string
	Pubdate      string
	Rating       int
	Languages    []string
	Identifiers  map[string]string
	Description  string // plain text
	Formats      map[string]string // format -> filename_without_ext
	RelativePath string // books.path
	LastModified string
}

// ParseCalibreLibrary loads all books from a Calibre library (read-only).
func ParseCalibreLibrary(libraryPath string) ([]*CalibreBook, error) {
	dbPath := filepath.Join(libraryPath, "metadata.db")
	uri := fmt.Sprintf("file:%s?mode=ro", dbPath)

	conn, err := sql.Open("sqlite3", uri)
	if err != nil {
		return nil, fmt.Errorf("open calibre db: %w", err)
	}
	defer conn.Close()

	return loadAllBooks(conn)
}

// GetBookFilePath resolves the absolute path for the best available format.
func GetBookFilePath(libraryPath string, book *CalibreBook, preferredFormats []string) (string, string) {
	if len(preferredFormats) == 0 {
		preferredFormats = []string{"EPUB", "PDF"}
	}

	for _, fmt := range preferredFormats {
		filenameBase, ok := book.Formats[fmt]
		if !ok {
			continue
		}
		filePath := filepath.Join(libraryPath, book.RelativePath, filenameBase+"."+strings.ToLower(fmt))
		if fileExists(filePath) {
			return filePath, strings.ToLower(fmt)
		}
	}
	return "", ""
}

func loadAllBooks(conn *sql.DB) ([]*CalibreBook, error) {
	rows, err := conn.Query("SELECT id, title, path, pubdate, last_modified FROM books")
	if err != nil {
		return nil, fmt.Errorf("query books: %w", err)
	}
	defer rows.Close()

	type bookRow struct {
		id           int64
		title        string
		path         string
		pubdate      string
		lastModified string
	}
	var bookRows []bookRow
	for rows.Next() {
		var b bookRow
		var pubdate, lastMod sql.NullString
		if err := rows.Scan(&b.id, &b.title, &b.path, &pubdate, &lastMod); err != nil {
			continue
		}
		if pubdate.Valid {
			b.pubdate = pubdate.String
		}
		if lastMod.Valid {
			b.lastModified = lastMod.String
		}
		bookRows = append(bookRows, b)
	}

	if len(bookRows) == 0 {
		return nil, nil
	}

	authorsMap := loadBookAuthors(conn)
	tagsMap := loadBookTags(conn)
	seriesMap := loadBookSeries(conn)
	publishersMap := loadBookPublishers(conn)
	ratingsMap := loadBookRatings(conn)
	languagesMap := loadBookLanguages(conn)
	identifiersMap := loadBookIdentifiers(conn)
	commentsMap := loadBookComments(conn)
	formatsMap := loadBookFormats(conn)

	books := make([]*CalibreBook, 0, len(bookRows))
	for _, r := range bookRows {
		description := ""
		if html, ok := commentsMap[r.id]; ok {
			description = HTMLToText(html)
		}

		series := ""
		seriesIndex := 0.0
		if si, ok := seriesMap[r.id]; ok {
			series = si.name
			seriesIndex = si.index
		}

		books = append(books, &CalibreBook{
			BookID:       r.id,
			Title:        r.title,
			Authors:      authorsMap[r.id],
			Tags:         tagsMap[r.id],
			Series:       series,
			SeriesIndex:  seriesIndex,
			Publisher:    publishersMap[r.id],
			Pubdate:      r.pubdate,
			Rating:       ratingsMap[r.id],
			Languages:    languagesMap[r.id],
			Identifiers:  identifiersMap[r.id],
			Description:  description,
			Formats:      formatsMap[r.id],
			RelativePath: r.path,
			LastModified: r.lastModified,
		})
	}

	slog.Info("loaded books from Calibre library", "count", len(books))
	return books, nil
}

func loadBookAuthors(conn *sql.DB) map[int64][]string {
	result := make(map[int64][]string)
	rows, err := conn.Query(
		"SELECT bal.book, a.name FROM books_authors_link bal JOIN authors a ON bal.author = a.id ORDER BY bal.book, bal.id",
	)
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var name string
		if rows.Scan(&bookID, &name) == nil {
			result[bookID] = append(result[bookID], name)
		}
	}
	return result
}

func loadBookTags(conn *sql.DB) map[int64][]string {
	result := make(map[int64][]string)
	rows, err := conn.Query(
		"SELECT btl.book, t.name FROM books_tags_link btl JOIN tags t ON btl.tag = t.id",
	)
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var name string
		if rows.Scan(&bookID, &name) == nil {
			result[bookID] = append(result[bookID], name)
		}
	}
	return result
}

type seriesInfo struct {
	name  string
	index float64
}

func loadBookSeries(conn *sql.DB) map[int64]seriesInfo {
	result := make(map[int64]seriesInfo)
	rows, err := conn.Query(
		"SELECT bsl.book, s.name, b.series_index FROM books_series_link bsl JOIN series s ON bsl.series = s.id JOIN books b ON bsl.book = b.id",
	)
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var name string
		var idx sql.NullFloat64
		if rows.Scan(&bookID, &name, &idx) == nil {
			si := seriesInfo{name: name}
			if idx.Valid {
				si.index = idx.Float64
			}
			result[bookID] = si
		}
	}
	return result
}

func loadBookPublishers(conn *sql.DB) map[int64]string {
	result := make(map[int64]string)
	rows, err := conn.Query(
		"SELECT bpl.book, p.name FROM books_publishers_link bpl JOIN publishers p ON bpl.publisher = p.id",
	)
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var name string
		if rows.Scan(&bookID, &name) == nil {
			result[bookID] = name
		}
	}
	return result
}

func loadBookRatings(conn *sql.DB) map[int64]int {
	result := make(map[int64]int)
	rows, err := conn.Query(
		"SELECT brl.book, r.rating FROM books_ratings_link brl JOIN ratings r ON brl.rating = r.id",
	)
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var rating int
		if rows.Scan(&bookID, &rating) == nil {
			result[bookID] = rating
		}
	}
	return result
}

func loadBookLanguages(conn *sql.DB) map[int64][]string {
	result := make(map[int64][]string)
	rows, err := conn.Query(
		"SELECT bll.book, l.lang_code FROM books_languages_link bll JOIN languages l ON bll.lang_code = l.id",
	)
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var lang string
		if rows.Scan(&bookID, &lang) == nil {
			result[bookID] = append(result[bookID], lang)
		}
	}
	return result
}

func loadBookIdentifiers(conn *sql.DB) map[int64]map[string]string {
	result := make(map[int64]map[string]string)
	rows, err := conn.Query("SELECT book, type, val FROM identifiers")
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var idType, val string
		if rows.Scan(&bookID, &idType, &val) == nil {
			if result[bookID] == nil {
				result[bookID] = make(map[string]string)
			}
			result[bookID][idType] = val
		}
	}
	return result
}

func loadBookComments(conn *sql.DB) map[int64]string {
	result := make(map[int64]string)
	rows, err := conn.Query("SELECT book, text FROM comments")
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var text string
		if rows.Scan(&bookID, &text) == nil {
			result[bookID] = text
		}
	}
	return result
}

func loadBookFormats(conn *sql.DB) map[int64]map[string]string {
	result := make(map[int64]map[string]string)
	rows, err := conn.Query("SELECT book, format, name FROM data")
	if err != nil {
		return result
	}
	defer rows.Close()
	for rows.Next() {
		var bookID int64
		var format, name string
		if rows.Scan(&bookID, &format, &name) == nil {
			if result[bookID] == nil {
				result[bookID] = make(map[string]string)
			}
			result[bookID][format] = name
		}
	}
	return result
}
